File: c++/Solution378.cpp
//Binary Search O(n * log(r - l)) O(1), n = matrix.size(), l = matrix[0][0], r = matrix[n - 1][n - 1]
class Solution378 {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        int l = matrix[0][0], r = matrix.back().back();
        while(l < r){
            int mid = l + (r - l) / 2;//防止負數, ex: (-5+(-4))/2=-4, -5+(-4-(-5))/2=-5
            if(k <= getRank(mid, matrix)){//可以再縮小
                r = mid;//保留大於的值
            }else{
                l = mid + 1;
            }
        }
        return r;
    }

    int getRank(int target, vector<vector<int>>& matrix){
        int i = matrix.size() - 1, j = 0, cnt = 0;
        while(i >= 0 && j <= matrix.size() - 1){
            if(matrix[i][j] <= target){
                cnt += i + 1;//加入一整列
                ++j;
            }else{
                --i;
            }
        }
        return cnt;
    }
};
/* rows and columns is sorted in ascending, 行列都是排序過

mid=7

1   2   3 __4|
2   4   6|  8
3   6 __7|  9
4   7|  10  12

*/

File: java/Solution378.java
package leetCode.java;

//Binary Search O(n * log(r - l)) O(1), n = matrix.length, l = matrix[0][0], r = matrix[n - 1][n - 1]
class Solution378 {
    public int kthSmallest(int[][] matrix, int k) {
        int n = matrix.length;
        int l = matrix[0][0], r = matrix[n - 1][n - 1];
        while(l < r){
            int mid = l + (r - l) / 2;//防止負數問題與更大容量
            if(k <= getRank(mid, matrix)){
                r = mid;
            }else{
                l = mid + 1;
            }
        }
        return r;
    }

    public int getRank(int target, int[][] matrix){
        int n = matrix.length;
        int cnt = 0, i = n - 1, j = 0;
        while(i >= 0 && j < n){
            if(target >= matrix[i][j]){
                cnt += i + 1;
                ++j;
            }else{
                --i;
            }
        }
        return cnt;
    }
}
/*
    vj
    1   2   3
    2   3   4
  i>3   4   5

 */